app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt

```diff
@@ -147,69 +151,57 @@ class SaleDatabase : SaleDatabase {
                 return future
             }
 
-            // FIXME ignoring page number for now
-            getQuery()
-                .limit(n.toLong())
-                .get()
-                .addOnSuccessListener { documents ->
-                    future.complete(documents.map { document ->
-                        snapshotToSale(document)
-                    })
-                }
-                .addOnFailureListener {
-                    future.completeExceptionally(
-                        DatabaseException("Query could not be completed")
-                    )
+            if (interests == null && title == null && isbn == null) { //In this case we should not look in the book database
+                return doQuery(paginateQuery(filterQuery(saleRef), n, page)).thenCompose { snapshotsToSales(it) }
+            } else {
+                val booksFuture = getBookQuery().getAll()
+                return booksFuture.thenCompose { books ->  //those are the books for which we want to find the sales
+                    val isbns = books.map {it.isbn}
+                    val isbnToBook = books.associateBy { it.isbn } //is used a cache to transform snapshots to Sales
+                    val saleQuery = saleRef.whereIn(SaleFields.BOOK_ISBN.fieldName, isbns)
+                    doQuery(paginateQuery(filterQuery(saleQuery), n, page)).thenCompose { snapshotsToSales(it,isbnToBook) }
                 }
+            }
 
             return future
         }
 
         override fun getCount(): CompletableFuture<Int> {
-            val future: CompletableFuture<Int> = CompletableFuture()
-
-            getQuery()
-                .get()
-                .addOnSuccessListener { documents ->
-                    future.complete(documents.fold(0){ acc, _ -> acc + 1 })
-                }
-                .addOnFailureListener {
-                    future.completeExceptionally(
-                        DatabaseException("Query count could not be completed")
-                    )
+            if (interests == null && title == null && isbn == null) { //In this case we should not look in the book database
+                return doQuery(filterQuery(saleRef)).thenApply { it.size() }
+            } else {
+                val booksFuture = getBookQuery().getAll()
+                return booksFuture.thenCompose { books ->  //those are the books for which we want to find the sales
+                    if (books.isEmpty()) return@thenCompose CompletableFuture.completedFuture(0)
+                    val isbns = books.map {it.isbn}
+                    val saleQuery = saleRef.whereIn(SaleFields.BOOK_ISBN.fieldName, isbns)
+                    doQuery(filterQuery(saleQuery)).thenApply { it.size() }
```

Comment:
maybe you can put this part in a private method to avoid code duplication

Reply:
I tried, but factoring it isn't really possible for getN(), getAll(), and getCount(). 
There is the ```.thenCompose { snapshotsToSales(it,isbnToBook) }``` to share a private method between getCount() and getN() because it uses ```isbnToBook```.